### PR TITLE
swarm: migrated overlay sim to rewrite

### PR DESCRIPTION
**Pull request by holisticode**
This PR makes the overlay simulation available to the network rewrite branch. 

It also newly introduces optional throttling for the server-side HTTP event stream of simulations, as with many nodes the simulation frontend may not be able to render in time before the next event comes in.
Also a few bug fixes.

### CODE REVIEW

**Comment by holisticode** [comment]
closes #207

File: p2p/simulations/http.go [hunk 1]
```diff
@@ -380,6 +381,13 @@ func (s *Server) ResetNetwork(w http.ResponseWriter, req *http.Request) {
 
 // StreamNetworkEvents streams network events as a server-sent-events stream
 func (s *Server) StreamNetworkEvents(w http.ResponseWriter, req *http.Request) {
+	var throttle <-chan time.Time
+	//When running a sim, the frontend may "choke to death" depending on params (node count)
+	//due to overwhelming event rate. It is therefore important to be able to throttle
+	//the rate at which the events are emitted, in order for the frontend to have time
+	//to terminate rendering before the next event arrives.
+	throttling := false
```
Comment (lmars) [inline p2p/simulations/http.go:389]: I don't really agree with this, can the events not be buffered and dropped client side? It seems that is where the issue is here.

File: p2p/simulations/mocker.go [hunk 2]
```diff
@@ -102,6 +102,11 @@ func startStop(net *Network, quit chan struct{}, nodeCount int) {
 func probabilistic(net *Network, quit chan struct{}, nodeCount int) {
 	nodes, err := connectNodesInRing(net, nodeCount)
 	if err != nil {
+		log.Error(fmt.Sprintf("%s", err))
```
Comment (lmars) [inline p2p/simulations/mocker.go]: This should be more informative, for example:

```
log.Error("error connecting nodes in a ring", "err", err)
```

File: p2p/simulations/mocker.go [hunk 3]
```diff
@@ -102,6 +102,11 @@ func startStop(net *Network, quit chan struct{}, nodeCount int) {
...
+		if _, ok := <-quit; ok == false {
+			//error may be due to abortion of mocking; so the quit channel is closed
+			return
+		}
```
Comment (lmars) [inline p2p/simulations/mocker.go]:
```go
select {
case <-quit:
	return
default:
	panic("...")
}
```

File: p2p/simulations/mocker.go [hunk 4]
```diff
@@ -167,7 +172,7 @@ func connectNodesInRing(net *Network, nodeCount int) ([]discover.NodeID, error)
 	for i := 0; i < nodeCount; i++ {
 		node, err := net.NewNode()
 		if err != nil {
-			log.Error("Error creating a node! %s", err)
+			log.Error(fmt.Sprintf("Error creating a node! %s", err))
```
Comment (lmars) [inline p2p/simulations/mocker.go]: I recommend putting the error as a separate key:

```go
log.Error("error creating node", "err", err)
```

File: p2p/simulations/http.go (hunk 1 above)
Comment (holisticode) [reply p2p/simulations/http.go:389]: Well excellent objection. My biggest argument is that it is much easier to do it here in go than in the frontend in javascript. Furthermore, the javascript client needs to subscribe to the event stream directly with the HTTP Server-side Eventstream API, thus it gets flooded with all the requests and would need to handle them. Javascript being singlethreaded, apart from more complex in the implementation, such an approach doesn't relieve the frontend from the load, which is the main reason to do this throttling in the first place. An effective way to do this client side would probably be building some kind of queue, which may imply additional processes and more moving pieces (I indeed started a first attempt with this in mind, but then ditched it).

In short - this is way easier, effective and efficient to do here in the backend, Also, it is optional and disabled by default, only if provided in the query string it will be activated. 

BTW the events are not being dropped but throttled.

File: p2p/simulations/mocker.go (hunk 2 above)
Comment (holisticode) [reply p2p/simulations/mocker.go]: Actually the line is not even needed, as `connectNodesInRing` already logs error events before returning an `error`

**Comment by zelig** [comment]
@holisticode status on this?

**Comment by holisticode** [comment]
Closed as needed to be PRd on `swarm-network-rewrite`
